### ISSUE TITLE
F# DataReader closed over Seq crashes with NullPointerException

### ISSUE THREAD

**Issue by giuliohome** [dotnet/fsharp#8897]
While C# `IDataReader` can be closed over a lazy `IEnumerable` correctly, F# `IDataReader` closed over a bugged lazy `Seq` crashes with `NullPointerException`

This is the perfect (no errors) C# reference code, from [my full repo](https://github.com/giuliohome/Donald/blob/master/TestFromCSharp/Giulio.cs), notice especially the following:

```
using (var DR = cmd.ExecuteReader())
{
    while (DR.Read())
    {
        yield return map(DR);
    }
}
```

**Repro steps**

Whilst the following is the exactly correspondent F# code that produces instead a `NullPointerException`:

```
let tranQuery sql param (map:System.Func<IDataReader, 'a>) tran =
    use cmd = newCommand sql param tran
    use rd = cmd.ExecuteReader()
    seq{ while rd.Read() do
                yield map.Invoke rd }
```

**Expected behavior**

F# should yield the **Seq** results like C# `IEnumerable` without NPE. 

**Actual behavior**

F# code throws a `NullPointerException`.

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=System.Data.SQLite
  StackTrace:
   at System.Data.SQLite.SQLiteDataReader.GetOrdinal(String name)
   at TestFromCSharp.Program.MyAuthor.FromReader(IDataReader rd) in E:\giulio-vs-so\Donald\TestFromCSharp\Program.cs:line 19
   at Donald.tranQuery@65.GenerateNext(IEnumerable`1& next) in E:\giulio-vs-so\Donald\src\Donald\Sql.fs:line 66
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seqcore.fs:line 371
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System-Collections-IEnumerator-MoveNext() in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seqcore.fs:line 403
   at TestFromCSharp.Program.Main(String[] args) in E:\giulio-vs-so\Donald\TestFromCSharp\Program.cs:line 43

```
**Known workarounds**

A known workaround is using an `array` or a `list` in F# inside the `use` binding of the `IDataReader`.

**Related information**

See the full code inside [my repo](https://github.com/giuliohome/Donald),

* Operating system Windows 64 bit
* .NET Runtime kind .NET STANDARD
* Editing Tools Visual Studio 2019

**Comment by zhuangzhuang** [comment]
try
```F#
let tranQuery sql param (map:System.Func<IDataReader, 'a>) tran =    
    seq{ 
        use cmd = newCommand sql param tran
        use rd = cmd.ExecuteReader()
        while rd.Read() do
                yield map.Invoke rd }
```

**Comment by giuliohome** [comment]
@zhuangzhuang same thing: NPE in F#, confirmed after such specific test done now, as per your request: indeed it should be the same for a syntax point of view.

**Comment by giuliohome** [comment]
I see, the relevant point is that F# disposes the transaction, not the datareader...

```
let query sql param map conn =
    use tran = beginTran conn
    let results = tranQuery sql param map tran 
```

**Comment by giuliohome** [comment]
But the corresponding C# with a `using` for a wrapping transaction is still OK...

**Comment by giuliohome** [comment]
So I can rewrite it all also as

```
let query sql param (map:System.Func<IDataReader, 'a>) (conn : IDbConnection) =
    use tran = conn.BeginTransaction()
    let results = //tranQuery sql param map tran
        seq {
        use cmd = tran.Connection.CreateCommand()
        cmd.CommandType <- CommandType.Text
        cmd.CommandText <- sql
        cmd.Transaction <- tran

        let createParam param = 
            let p = cmd.CreateParameter()
            p.ParameterName <- param.Name
            p.Value <- param.Value
            cmd.Parameters.Add(p) |> ignore
       
        param |> Seq.iter createParam

        use rd = cmd.ExecuteReader()
        // https://github.com/dotnet/fsharp/issues/8897
        // workaround array or list
        while rd.Read() do
                    yield map.Invoke rd }

    tran.Commit()
    results
```

that still throws (only in F#, not in the corresponding C# as per my previous comment)

```
System.ObjectDisposedException
  HResult=0x80131622
  Message=Cannot access a disposed object.
Object name: 'SQLiteTransactionBase'.
  Source=System.Data.SQLite
  StackTrace:
   at System.Data.SQLite.SQLiteTransactionBase.CheckDisposed()
   at System.Data.SQLite.SQLiteTransactionBase.get_DbConnection()
   at System.Data.Common.DbTransaction.System.Data.IDbTransaction.get_Connection()
   at Donald.results@128.GenerateNext(IEnumerable`1& next) in E:\giulio-vs-so\Donald\src\Donald\Sql.fs:line 129
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seqcore.fs:line 371
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System-Collections-IEnumerator-MoveNext() in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seqcore.fs:line 403
   at TestFromCSharp.Program.Main(String[] args) in E:\giulio-vs-so\Donald\TestFromCSharp\Program.cs:line 43
```

**Comment by giuliohome** [comment]
Or, as stated at start, a NPE (*) for the following similar version (they are all equivalent syntax variants in F#)

```
let query sql param (map:System.Func<IDataReader, 'a>) (conn : IDbConnection) =
    use tran = conn.BeginTransaction()
    use cmd = tran.Connection.CreateCommand()
    cmd.CommandType <- CommandType.Text
    cmd.CommandText <- sql
    cmd.Transaction <- tran

    let createParam param = 
        let p = cmd.CreateParameter()
        p.ParameterName <- param.Name
        p.Value <- param.Value
        cmd.Parameters.Add(p) |> ignore

    param |> Seq.iter createParam

    use rd = cmd.ExecuteReader()
    let results = //tranQuery sql param map tran
        seq {
        // https://github.com/dotnet/fsharp/issues/8897
        // workaround array or list
        while rd.Read() do
                    yield map.Invoke rd }

    tran.Commit()
    results
```

(*)

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=System.Data.SQLite
  StackTrace:
   at System.Data.SQLite.SQLiteDataReader.GetOrdinal(String name)
   at TestFromCSharp.Program.MyAuthor.FromReader(IDataReader rd) in E:\giulio-vs-so\Donald\TestFromCSharp\Program.cs:line 19
   at Donald.results@144.GenerateNext(IEnumerable`1& next) in E:\giulio-vs-so\Donald\src\Donald\Sql.fs:line 145
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seqcore.fs:line 371
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System-Collections-IEnumerator-MoveNext() in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seqcore.fs:line 403
   at TestFromCSharp.Program.Main(String[] args) in E:\giulio-vs-so\Donald\TestFromCSharp\Program.cs:line 43
```

**Comment by giuliohome** [comment]
@zhuangzhuang thank you
Finally, this works without any NPE in F#
But I assume it is a workaround, correct?

```
let query sql param (map:System.Func<IDataReader, 'a>) (conn : IDbConnection) =
    //let results = //tranQuery sql param map tran
        seq {
        
        use tran = conn.BeginTransaction()
        use cmd = tran.Connection.CreateCommand()
        cmd.CommandType <- CommandType.Text
        cmd.CommandText <- sql
        cmd.Transaction <- tran

        let createParam param = 
            let p = cmd.CreateParameter()
            p.ParameterName <- param.Name
            p.Value <- param.Value
            cmd.Parameters.Add(p) |> ignore

        param |> Seq.iter createParam

        use rd = cmd.ExecuteReader()
        // https://github.com/dotnet/fsharp/issues/8897
        // workaround array or list
        while rd.Read() do
                    yield map.Invoke rd 
        tran.Commit()            
        }
```

**Comment by giuliohome** [comment]
Well, I admit that enumeration is need also in C# when a transactional wrapper is added, so we can close this since the situation for C# is equivalent or you can keep this open if you are able to make a better enhancement in F#

**Comment by giuliohome** [comment]
Yes, this works fine

```
let query sql param map conn =
    seq {
    use tran = beginTran conn
    yield! tranQuery sql param map tran
    commitTran tran
    }
```
    
I'd rather prefer avoiding the double seq wrapper, but the yield bang is cool enough and I bet you'll never develop the enhancement I'd auggest

**Comment by giuliohome** [comment]
> Yes, this works fine
> 
> ```
> let query sql param map conn =
>     seq {
>     use tran = beginTran conn
>     yield! tranQuery sql param map tran
>     commitTran tran
>     }
> ```
> 
> I'd rather prefer avoiding the double seq wrapper, but the yield bang is cool enough and I bet you'd never develop the enhancement I'd suggest

Notice that however try/with **can't** be used with sequence, so the issue is closed **without** a perfect solution (a true transactional and lazy wrapper can't be really written in pure functional style), but only on the bona fide assumption that you wouldn't fix it anyway.

**Comment by pimbrouwers** [comment]
In my experience seq computation expressions, which is what you're using here just don't play nicely in this particular area. Why not just eagerly evaluate the collection? What benefit are you REALLY deriving here from lazily evaluating this collection?

**Comment by giuliohome** [comment]
@pimbrouwers 
Hi
The benefit of seq is that it is exactly the innumerable of c# so it is great for an api. That can be easily solved in a fictitious way after an eager evaluation. The other benefit, if you have worked with linq2db is that you can dynamically add other conditions (like a where) after the call and change the db sql, not just the results in memory. Anyway I finally have a working code that is lazy evaluated. I was disappointed for not being able to try/with but, since it's lazy, it can be delegated to the caller for a reason.
You asked about the Close on twitter but at the moment I think you are right, it is not needed (I've a slight doubt maybe for odbc but I'm no longer using it).
The only think I'd suggest (not related to this issue) is an async version. Here I'm a bit biased towards c# async await task... but I know that also F# is converging toward a task CE...

**Comment by pimbrouwers** [comment]
@giuliohome 

I'll add the close, as I'm more interested in being fully covered than anything else. Want to submit an issue? Or even better yet a PR?

I thought long and and hard about async methods, as I know many people look for them. But in my recent experiments, I've actually come to learn that async + ADO comes at a steep cost. We recently removed all of the async db-related calls, preserving async throughout the rest of the app-stack. The performance gain was more than measureable. 

I believe the problem has to do with connection pooling, and available workers within the database engine itself. With async, since threads will "go and get more work", in database intensive applications this often means another latch into the database. So you end up flooding your db with requests to begin work.

That said, I am not opposed to adding these methods in if people find them useful. I've created an enhancement [issue](https://github.com/pimbrouwers/Donald/issues/2) for us to discuss.
